### PR TITLE
FEATURE: Polymorphic bookmarks pt. 1 (CRUD)

### PR DESCRIPTION
**Pull request by martin-brennan**
This commit introduces a new use_polymorphic_bookmarks site setting
that is default false and hidden, that will be used to help continuous
development of polymorphic bookmarks. This setting **should not** be
enabled anywhere in production yet, it is purely for local development.

This commit uses the setting to enable create/update/delete actions
for polymorphic bookmarks on the server and client side. The bookmark
interactions on topics/posts are all usable. Listing, searching,
sending bookmark reminders, and other edge cases will be handled
in subsequent PRs.

Comprehensive UI tests will be added in the final PR -- we already
have them for regular bookmarks, so it will just be a matter of
changing them to be for polymorphic bookmarks.

### CODE REVIEW

**Comment by martin-brennan** [comment]
Failing spec is unrelated and will be fixed by https://github.com/discourse/discourse/pull/16309

File: app/assets/javascripts/discourse/app/controllers/bookmark.js [hunk 1]
```diff
@@ -10,28 +11,54 @@ export function openBookmarkModal(
...
+        if (bookmark.for_topic) {
+          return I18n.t(
+            bookmark.id
+              ? "post.bookmarks.edit_for_topic"
+              : "post.bookmarks.create_for_topic"
+          );
+        }
+        return I18n.t(
+          bookmark.id ? "post.bookmarks.edit" : "post.bookmarks.create"
+        );
```
Comment (tgxworld) [inline app/assets/javascripts/discourse/app/controllers/bookmark.js:38]:
```suggestion
      if (options.use_polymorphic_bookmarks) {
        return I18n.t(
          bookmark.id ? "bookmarks.edit_for" : "bookmarks.create_for",
          {
            type: bookmark.bookmarkable_type,
          }
        );
      } else if (bookmark.for_topic) {
        return I18n.t(
          bookmark.id
            ? "post.bookmarks.edit_for_topic"
            : "post.bookmarks.create_for_topic"
        );
      } else {
        return I18n.t(
          bookmark.id ? "post.bookmarks.edit" : "post.bookmarks.create"
        );
      }
```

Comment (tgxworld) [reply app/assets/javascripts/discourse/app/controllers/bookmark.js:38]: The additional nesting in `else` made the code harder to read.

File: app/models/bookmark.rb [hunk 2]
```diff
@@ -24,19 +24,37 @@ def self.auto_delete_preferences
...
   validate :unique_per_post_for_user,
     on: [:create, :update],
     if: Proc.new { |b| b.will_save_change_to_post_id? || b.will_save_change_to_user_id? }
 
+  # TODO (martin) [POLYBOOK] Not relevant once polymorphic bookmarks are implemented.
   validate :for_topic_must_use_first_post,
     on: [:create, :update],
     if: Proc.new { |b| b.will_save_change_to_post_id? || b.will_save_change_to_for_topic? }
 
+  validate :unique_per_bookmarkable,
```
Comment (tgxworld) [inline app/models/bookmark.rb:37]: We will need a validation to ensure that `bookmarkable_id` and `bookmarkable_type` are both present if one is present too.

File: lib/bookmark_manager.rb [hunk 3]
```diff
@@ -5,6 +5,33 @@ class BookmarkManager
...
+
+  # TODO (martin) [POLYBOOK] This will be used in place of #create once
+  # polymorphic bookmarks are implemented.
+  def create_for(bookmarkable_id:, bookmarkable_type:, name: nil, reminder_at: nil, options: {})
+    raise NotImplementedError if !SiteSetting.use_polymorphic_bookmarks
+
+    bookmarkable = bookmarkable_type.constantize.find_by(id: bookmarkable_id)
+    self.send("validate_bookmarkable_#{bookmarkable_type.downcase}", bookmarkable)
+
+    bookmark = Bookmark.create(
```
Comment (tgxworld) [inline lib/bookmark_manager.rb:19]: Do we need the same guardian checks as `BookmarkManager.create`?

File: lib/bookmark_manager.rb [hunk 4]
```diff
@@ -157,4 +181,31 @@ def update_topic_user_bookmarked(topic, opts = {})
...
+  def after_create_bookmarkable_post(post, opts = {})
+    update_topic_user_bookmarked(post.topic, opts)
+  end
+
+  def after_create_bookmarkable_topic(topic, opts = {})
+    update_topic_user_bookmarked(topic, opts)
+  end
+
+  def after_destroy_bookmarkable_post(bookmark)
+    { topic_bookmarked: update_topic_user_bookmarked(bookmark.bookmarkable.topic) }
```
Comment (tgxworld) [inline lib/bookmark_manager.rb:194]: I feel like it isn't this method's responsibility to know what the return object should be for another method.

File: lib/topic_view.rb [hunk 5]
```diff
@@ -413,9 +413,15 @@ def has_bookmarks?
   end
 
   def bookmarks
-    @bookmarks ||= @topic.bookmarks.where(user: @user).joins(:topic).select(
-      :id, :post_id, "topics.id AS topic_id", :for_topic, :reminder_at, :name, :auto_delete_preference
-    )
+    if SiteSetting.use_polymorphic_bookmarks
+      @bookmarks ||= Bookmark.for_user_in_topic(@user&.id, @topic.id).select(
```
Comment (tgxworld) [inline lib/topic_view.rb:417]: I don't think we need to execute a query when for an anon user. We can just return an empty array early if `@user&.id` is `nil`

File: lib/topic_view.rb [hunk 6]
```diff
@@ -512,8 +518,8 @@ def links
     @links ||= TopicLink.topic_map(@guardian, @topic.id)
   end
 
-  def user_post_bookmarks
-    @user_post_bookmarks ||= @topic.bookmarks.where(user: @user)
+  def user_bookmarks
+    @user_bookmarks ||= Bookmark.for_user_in_topic(@user&.id, @topic.id)
```
Comment (tgxworld) [inline lib/topic_view.rb:522]: I don't think we need to execute a query when for an anon user. We can just return an empty array early if `@user&.id` is `nil`

Comment (tgxworld) [reply lib/topic_view.rb:522]: Just curious, what is the difference between `user_bookmarks` and `bookmarks`?

File: spec/lib/bookmark_manager_spec.rb [hunk 7]
```diff
@@ -334,4 +379,127 @@ def notifications_for_user
...
+
+  describe "#create_for (use_polymorphic_bookmarks)" do
+    before do
+      SiteSetting.use_polymorphic_bookmarks = true
+    end
+
+    it "allows creating a bookmark for the topic and for the first post" do
+      subject.create_for(bookmarkable_id: post.topic_id, bookmarkable_type: "Topic", name: name)
+      bookmark = Bookmark.find_by(user: user, bookmarkable: post.topic)
+      expect(bookmark).not_to eq(nil)
```
Comment (tgxworld) [inline spec/lib/bookmark_manager_spec.rb:391]:
```suggestion
      expect(bookmark.present?).to eq(true)
```
Eliminates the double negative.

File: spec/lib/bookmark_manager_spec.rb [hunk 8]
```diff
@@ -334,4 +379,127 @@ def notifications_for_user
...
+
+      subject.create_for(bookmarkable_id: post.id, bookmarkable_type: "Post", name: name)
+      bookmark = Bookmark.find_by(user: user, bookmarkable: post)
+      expect(bookmark).not_to eq(nil)
+    end
+
+    it "when topic is deleted it raises invalid access from guardian check" do
```
Comment (tgxworld) [inline spec/lib/bookmark_manager_spec.rb:398]: Minor but I think guardian is an internal implementation detail here which we do not have to mention.

File: lib/topic_view.rb (hunk 6 above)
Comment (martin-brennan) [reply lib/topic_view.rb:522]: Oh yeah you're right...will remove one. And as for the query for anon, I will return an empty array, might just need some extra checks higher up the chain or some optional param for ordering, because I think in one place we order this

File: lib/bookmark_manager.rb (hunk 4 above)
Comment (martin-brennan) [reply lib/bookmark_manager.rb:194]: Hmm I see what you mean, I guess we can just remove this return value in the manager completely and just query it elsewhere from whatever is destroying the bookmark (the controller)

File: lib/topic_view.rb (hunk 6 above)
Comment (martin-brennan) [reply lib/topic_view.rb:522]: Never mind about the query thing, I was reading the wrong method! Did the rename though, and we don't have to worry about anon because the `has_bookmarks?` method protects us there :+1:

**Comment by martin-brennan** [comment]
@tgxworld This is ready for review again :+1:

**Review by tgxworld — APPROVED** [review]
Looks good. We do not have much client side tests now for the new site setting but that will be resolved once we fully switch to the polymorphic columns.

**Comment by martin-brennan** [comment]
> Looks good. We do not have much client side tests now for the new site setting but that will be resolved once we fully switch to the polymorphic columns.

@tgxworld yeah definitely will add them in later, I figure there is not much point for now with all the churn that will go on :+1:
